activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -1829,6 +1839,34 @@ def check_if_method_has_arguments!(method_name, args, message = nil)
         end
       end
 
+      def process_select_args(fields)
+        fields.map do |field|
+          if field.is_a?(Hash)
+            transform_select_hash_values(field)
+          else
+            field
+          end
+        end.flatten
+      end
+
+      def transform_select_hash_values(fields)
+        fields.each_with_object([]) do |(key, columns_aliases), result|
+          case columns_aliases
+          when Hash
+            columns_aliases.each do |column, column_alias|
+              result << "#{key}.#{column} as #{column_alias}"
+            end
+          when Array
+            columns_aliases.map do |column|
+              result << "#{key}.#{column}"
+            end
+          when String, Symbol
+            result << "#{key} as #{columns_aliases}"
```

Comment:
I'd need to consult with other members, but I think at least part of these should use Arel rather than generate strings, like in @simi's PR: https://github.com/rails/rails/pull/39874